File: August 2024 GFG SOLUTION/August-02.py
class Solution:
    def editDistance(self, str1, str2):
        m = len(str1)
        n = len(str2)
        
        # Create a 2D array to store minimum operations
        dp = [[0] * (n + 1) for _ in range(m + 1)]
        
        # Initialize dp array for base cases
        for i in range(m + 1):
            dp[i][0] = i  # Operations needed to convert str1[:i] to an empty string
        
        for j in range(n + 1):
            dp[0][j] = j  # Operations needed to convert an empty string to str2[:j]
        
        # Fill the dp array
        for i in range(1, m + 1):
            for j in range(1, n + 1):
                if str1[i - 1] == str2[j - 1]:
                    dp[i][j] = dp[i - 1][j - 1]  # Characters match, no new operation needed
                else:
                    dp[i][j] = 1 + min(dp[i - 1][j],     # Deletion
                                       dp[i][j - 1],     # Insertion
                                       dp[i - 1][j - 1]) # Replacement
        
        return dp[m][n]
    
#{
#Driver Code Starts
if __name__ == '__main__' :
    T - int(inout())
    for i in range(T) :
        s,t = input().split()
        ob = Solution()
        ans = ob.editDistance(s,t)
        print(ans)
        
# } Driver Code Ends

File: August 2024 GFG SOLUTION/August-02.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends

class Solution {
  public:
    int editDistance(string s, string t) {
        int m = s.size();
        int n = t.size();
        vector<vector<int>> dp(m + 1, vector<int>(n + 1));

        for (int i = 0; i <= m; ++i) {
            for (int j = 0; j <= n; ++j) {
                if (i == 0)
                    dp[i][j] = j;
                else if (j == 0)
                    dp[i][j] = i;
                else if (s[i - 1] == t[j - 1])
                    dp[i][j] = dp[i - 1][j - 1];
                else
                    dp[i][j] = 1 + min({dp[i][j - 1], dp[i - 1][j], dp[i - 1][j - 1]});
            }
        }
        return dp[m][n];
    }
};


//{ Driver Code Starts.
int main() {
    int T;
    cin >> T;
    while (T--) {
        string s, t;
        cin >> s >> t;
        Solution ob;
        int ans = ob.editDistance(s, t);
        cout << ans << "\n";
    }
    return 0;
}

// } Driver Code Ends
